tomcat/tomcat-jta/src/main/java/org/jboss/narayana/tomcat/jta/TransactionalDataSourceFactory.java

```diff
@@ -108,26 +109,29 @@ public boolean initialise(String p) throws Exception {
                     }
 
                     @Override
-                    public XAResource[] getXAResources() throws Exception {
-                        try {
-                            String user = properties.getProperty(PROP_USERNAME);
-                            String password = properties.getProperty(PROP_PASSWORD);
-
-                            if (user != null && password != null) {
-                                return new XAResource[]{xaDataSource.getXAConnection(user, password).getXAResource()};
-                            } else {
-                                return new XAResource[]{xaDataSource.getXAConnection().getXAResource()};
+                    public synchronized XAResource[] getXAResources() throws Exception {
+                        synchronized (lock) {
+                            initialiseConnection();
+                            try {
+                                String user = properties.getProperty(PROP_USERNAME);
```

Comment:
this is a bit strange to gather value for `user`/`password` and not using it further in the `try-block` and then the block for `user != null` is the same of functionality as if `user == null`